--- a/qsdk/qca/feeds/shortcut-fe/shortcut-fe/Makefile
+++ b/qsdk/qca/feeds/shortcut-fe/shortcut-fe/Makefile
@@ -32,7 +32,7 @@ define KernelPackage/shortcut-fe
   SECTION:=kernel
   CATEGORY:=Kernel modules
   SUBMENU:=Network Support
-  DEPENDS:=
+  DEPENDS:=libnetfilter-conntrack
   TITLE:=Kernel driver for SFE
   FILES:=$(PKG_BUILD_DIR)/shortcut-fe/shortcut-fe.ko $(PKG_BUILD_DIR)/shortcut-fe/shortcut-fe-ipv6.ko
   KCONFIG:=CONFIG_NF_CONNTRACK_EVENTS=y
