diff --git a/ecm_db/ecm_db_iface.c b/ecm_db/ecm_db_iface.c
index 5ae31c1..22ffb5f 100644
--- a/ecm_db/ecm_db_iface.c
+++ b/ecm_db/ecm_db_iface.c
@@ -1585,7 +1585,7 @@ EXPORT_SYMBOL(ecm_db_iface_find_and_ref_by_interface_identifier);
  * ecm_db_iface_ifidx_find_and_ref_ethernet()
  *	Return an interface based on a MAC address and interface hlos interface identifier
  */
-struct ecm_db_iface_instance *ecm_db_iface_ifidx_find_and_ref_ethernet(uint8_t *address, int32_t ifidx)
+struct ecm_db_iface_instance *ecm_db_iface_ifidx_find_and_ref_ethernet(uint8_t *address, int32_t ifidx, int32_t ae_interface_num)
 {
 	ecm_db_iface_hash_t hash_index;
 	struct ecm_db_iface_instance *ii;
@@ -1605,7 +1605,8 @@ struct ecm_db_iface_instance *ecm_db_iface_ifidx_find_and_ref_ethernet(uint8_t *
 	while (ii) {
 		if ((ii->type != ECM_DB_IFACE_TYPE_ETHERNET)
 		    || memcmp(ii->type_info.ethernet.address, address, ETH_ALEN)
-		    || ii->interface_identifier != ifidx) {
+		    || (ii->interface_identifier != ifidx)
+		    || (ii->ae_interface_identifier != ae_interface_num)) {
 			ii = ii->hash_next;
 			continue;
 		}
diff --git a/ecm_db/ecm_db_iface.h b/ecm_db/ecm_db_iface.h
index 40a5a51..465594d 100644
--- a/ecm_db/ecm_db_iface.h
+++ b/ecm_db/ecm_db_iface.h
@@ -192,7 +192,7 @@ void ecm_db_iface_vlan_info_get(struct ecm_db_iface_instance *ii,
 #endif
 
 struct ecm_db_iface_instance *ecm_db_iface_find_and_ref_by_interface_identifier(int32_t interface_id);
-struct ecm_db_iface_instance *ecm_db_iface_ifidx_find_and_ref_ethernet(uint8_t *address, int32_t idx);
+struct ecm_db_iface_instance *ecm_db_iface_ifidx_find_and_ref_ethernet(uint8_t *address, int32_t idx, int32_t ae_interface_num);
 
 #ifdef ECM_INTERFACE_RAWIP_ENABLE
 struct ecm_db_iface_instance *ecm_db_iface_find_and_ref_rawip(uint8_t *address);
diff --git a/ecm_interface.c b/ecm_interface.c
index ce5a391..18e9417 100644
--- a/ecm_interface.c
+++ b/ecm_interface.c
@@ -1819,14 +1819,10 @@ static struct ecm_db_iface_instance *ecm_interface_ethernet_interface_establish(
 	/*
 	 * Locate the iface
 	 */
-	ii = ecm_db_iface_ifidx_find_and_ref_ethernet(type_info->address, dev_interface_num);
+	ii = ecm_db_iface_ifidx_find_and_ref_ethernet(type_info->address, dev_interface_num, ae_interface_num);
 
 	if (ii) {
 		DEBUG_TRACE("%px: iface established\n", ii);
-		/*
-		 * Update the accel engine interface identifier, just in case it was changed.
-		 */
-		ecm_db_iface_ae_interface_identifier_set(ii, ae_interface_num);
 		return ii;
 	}
 
@@ -1843,7 +1839,7 @@ static struct ecm_db_iface_instance *ecm_interface_ethernet_interface_establish(
 	 * Add iface into the database, atomically to avoid races creating the same thing
 	 */
 	spin_lock_bh(&ecm_interface_lock);
-	ii = ecm_db_iface_ifidx_find_and_ref_ethernet(type_info->address, dev_interface_num);
+	ii = ecm_db_iface_ifidx_find_and_ref_ethernet(type_info->address, dev_interface_num, ae_interface_num);
 	if (ii) {
 		spin_unlock_bh(&ecm_interface_lock);
 		ecm_db_iface_deref(nii);
